README
\documentclass[a4paper]{article}
\usepackage{url}
\title{mlgt}
\author{Dave T. Gerrard}

\begin{document}
\maketitle

\begin{abstract}
Processing and analysis of high throughput (Roche 454) sequences generated from multiple loci and multiple biological samples. Sequences are assigned to their locus and sample of origin, aligned and trimmed. Where possible, genotypes are called and variants mapped to known alleles.
\end{abstract}

\section{Introduction}

The purpose of mlgt is to genotype multiple loci from multiple samples, where the sequence data is a single output file from a high throughput sequencing machine (e.g. the Roche 454 system).  Of course, the samples, and markers need to have been properly prepared (amplified and tagged) prior to sequencing so that they can be sorted out again. 

The data is expected to be full length sequences from mixed PCR amplicons. The amplicons from different samples will have been barcoded with unique end sequences (barcodes, MIDs) and these sequences need to be provided by the user. A set of reference amplicon sequences (markers) are also required, one for each amplicon.

\textbf{Definitions}
\begin{description}
\item[allele] Top variants called by mlgt OR alleles from an external source.
\item[amplicon] The sequence segment amplified in PCR. 
\item[barode/MID] Short sequence tags ligated to ends of amplicons. 
\item[marker] A reference sequence against which all variants are aligned.
\item[sample] The biological samples. 
\end{description}

A note on markers. You should use a sequence that is contained within the expected amplicon. I recommend NOT including the primer sequence as this is constrained by the primers used and may not reflect the true sequence of the samples. If you wish to compare your variants with alleles from an external source (e.g. the HLA database) then your marker sequence should be precisely bounded by the sequence that is in BOTH your amplicon and the allele sequences (e.g. the exon only part is good). 

Currently, mlgt makes no checks for similarity across markers, but it does assign all sequences to the best possible marker, even if that marker was never amplified in your dataset. This is particularly an issue with HLA datasets where supposedly distinct loci have very similar sequences.

\section{Installation}

mlgt runs in R version 2.13 or greater \footnote[1]{It uses long variable names (\textgreater 256 bytes) only implemented since R 2.13.}.  mlgt depends on another R package (\emph{seqinr}) and several external applications. These must all be installed and working for mlgt to work.

To assign sequences and retrieve specific sequences, mlgt makes use of the NCBI programs \emph{formatdb}, \emph{blastall} and \emph{fastacommand}, which are available here: \url{ftp://ftp.ncbi.nih.gov/blast/executables/release/2.2.24/}

To align sequence variants, mlgt uses \emph{MUSCLE}, which is available here: \url{http://www.drive5.com/muscle/downloads.htm}

Download the correct versions for your system and follow the installation instructions (if any). Make a note of the installation directories or, even better, specify where you want the programs to be installed.

On my machine the formatdb program is in ``C:/Users/Public/Apps/Blast/bin/formatdb.exe'' and MUSCLE is in ``C:/Users/Public/Apps/Muscle/muscle3.8.31\_i86win32.exe''

The R package \emph{seqinr} should be installed from within R. You can do this from the packages menu or using this command:-

<<label=Install seqinr,eval=FALSE>>=
install.packages("seqinr")
@

\paragraph{}

To install mlgt from within R you can do this from the packages menu (Install packages from local zip files) or using this command giving the full path to the package zip archive.:- 
<<label=Install mlgt,eval=FALSE>>=
install.packages("mlgt_0.1.zip", repos=NULL)
@

\section{Using mlgt}

Once installation is complete, you can begin to use mlgt. Load the library.

<<label=Load library>>=
library(mlgt)
@

You will also need to specify the locations of the auxillary programs.
<<label=Set paths to external apps>>=
formatdbPath <- "C:/Users/Public/Apps/Blast/bin/formatdb.exe"
fastacmdPath <- "C:/Users/Public/Apps/Blast/bin/fastacmd.exe"
blastAllPath <- "C:/Users/Public/Apps/Blast/bin/blastall.exe"
musclePath <- "C:/Users/Public/Apps/Muscle/muscle3.8.31_i86win32.exe"
@

\subsection{Prepare the analysis}

Start each analysis in a clean directory, perhaps named for the sequencing run and nested within a folder of all runs for the project.

<<label=Set working directory,eval=FALSE>>=
analysisDir <-  "C:/Users/me/genoProject1/run1/analysis/"
setwd(analysisDir)
@



You will need to create some variables to describe your sequencing run. You need a named list of the MIDs/barcodes used to mark each end of the amplicons and a list of samples. The easiest way to get this is from a fasta file containing the barcode sequences with each sequence annotated with the sample name you will use. Example data are in the /data sub-directory of the mlgt package installation directory and can be found using \code{system.file()}.

<<label=Describe run,keep.source=TRUE>>=
system.file("data/namedBarcodes.fasta", package="mlgt")

# Load MIDs used to mark samples
fTagList <- read.fasta(system.file("data/namedBarcodes.fasta", package="mlgt"), 
			as.string=T) 
# here we're using the same tags at both ends of the amplicons.
rTagList <- fTagList
#The names of the samples
sampleList <- names(fTagList)
# Load the marker sequences. 
myMarkerList <- read.fasta(system.file("data/HLA_namedMarkers.fasta", package="mlgt"),
			as.string=T)	
		
# The fasta file of sequence reads
inputDataFile <- system.file("data/sampleSequences.fasta", package="mlgt")
@

Inspect what is stored in each variable by typing its name.

Now you can create an object of class \emph{mlgtDesign} to hold all this information. Give the names of the variables you have just created for the marker list, the sample list and the MIDs. Also give a project name and a name for this run; this will help to identify the source of this object later on. 

<<label=Prepare run,keep.source=TRUE>>=
# Creates object to store run settings
my.mlgt.Design <- prepareMlgtRun(projectName="myProject", 
				runName="myRun", samples=sampleList, 
				markers=myMarkerList, fTags=fTagList, 
				rTags=rTagList, inputFastaFile=inputDataFile, 
				overwrite="yes")
@

As this object is created, multiple BLAST databases are also created in the working directory and all the input sequences are BLASTed against the databases. These BLAST results are used to assign sequences to markers and samples. 

It might be instructive to see how many sequences in your dataset are being assigned to each marker, especially if the marker list includes sequences which were meant NOT to be targetted by your primers.

<<label=Inspect BLAST results>>=
# inspect BLAST results for a specific marker
thisMarker <- "DPA1_E2"
topHits <- getTopBlastHits(my.mlgt.Design@markerBlastResults)
#inspectBlastResults(topHits, thisMarker)
@

\begin{center}
<<label=plot_BLAST_results,fig=TRUE, echo=TRUE>>=
inspectBlastResults(topHits, thisMarker)
@
\end{center}


CHANGELOG.txt
mlgt changes

Ver 0.18

Altered mlgt() to allow different 5-prime and 3-prime MIDs.

Ver 0.17

Added numbAllelesToMatch parameter to callGenotypes() to allow matching of more than 2 alleles.

Added name and sequence checks on sample, marker and barcodes. All barcodes must match in length.

Ver 0.16

Changes to setMethod documentation to prevent warnings in CRAN.

Removed HLA/IMGT sample data. Replaced with download in README.


Ver 0.15

Fixed bug in callGenotypes that was not adding 'NA' columns using approxMatch=TRUE

Corrected location of example data.


Ver 0.14

Fixed bug in alignReport causing error when only 1 or 0 sequences.

Added copyIfSpaces() to copy original sequence files if their path contains spaces. 

Added combineMlgtResults() function to combine mlgt results. Especially useful if runnning mlgt in parallel but could also be used to combine results across runs.

Added 'filePrefix' parameter to writeGenotypeCallsToFile, allows users to distinguish sets of output tables.

Reworked callGenotypes() to allow user specification of custom methods. The previous default method callGenotypes.table() is now not aliased and is named callGenotypes.default(). Removed setGeneric call for callGenotypes() as there is now only one primary method.

Altered callGenotypes.default() to also check that the third most abundant variant accounts for  less than 10% of used sequences [maxPropVarThree=0.1].

Reworked getSubseqsTable to use [maxVarsToAlign=30] most common variants only when greater than a threshold [minTotalCount=500] number of raw sequences. This function now also returns an object of class 'varCount' which includes the table of variants as well as summary counts of variants and total reads. The result sub.alignments contain fewer raw variants and are much tidier. Updated mlgt() to accept the new varCount object and added new statistics as new columns in the table that appears in markerSampleList.

Reworked mlgt()/getSubseqsTable() to include errorCorrection [errorCorrect=TRUE].

Reworked mlgt()/getSubseqsTable() to include a length filter on raw variants [minLength=70].

Added similarity allele matching to callGenotypes [approxMatching=TRUE]. This uses a new internal function makeVarAlleleBlastMap() and maps variants to allele by greatest BLAST similarity. 

Added 'userAlignment' parameter to createKnownAlleleList() to allow user to supply alignments with marker pre-aligned to known alleles. 

Ver 0.13

mlgtResult now retains all aligned variants. Enables generation of alignment reports with goal of implementing error correction.

Added alignReport() function to give tabular and graphical reports on sample/marker alignments. Focused on site-specific minor allele frequency.

Added errorCorrect() function to create new mlgtResult object by correct low frequency variants, site-by-site.

Added dumpVariants() as a first method to output variants as fasta sequences. Overlaps a bit with dumpVariantMap.mlgtResult

Ver 0.12

Added checks for whitespace in paths. Terminates prepareMlgtRun if inputFastaFile has whitespace. Could not get System(formatdb) to work with them (even with multiple escapes). 

Replaced user specification of blastallPath, musclePath etc with environment variables BLASTALL_PATH, MUSCLE_PATH etc, which must be set by Sys.setenv. Added checks to prepareMlgtRun that they are set. Original variables still used internally. 


skeletonMLGT.R


### TO DO

# Load reference alleles from DB/flatfile
# Construct 'comparable amplicon tree' from target markers and reference alleles. 
# 	problematic if differences between alleles are indels. 
# 	What about missing intronic sequence? Need to know source of reference allele. 
# 	Perhaps also need to mark projects amplicon as genomic or mRNA. 
# Compare with existing alleles
# checks for uniqueness of raw read IDs
# checks for uniqueness of marker IDs, MIDs
# Reduce I/O. Pipe or stdin/stdout from auxillary progs.
#	command <- "C:/Users/Public/Apps/Blast/bin/fastacmd.exe -p F -I  -d inputSeqs"
#	test <- readLines(pipe(command))
# Or set up proper MPI?
# Functions.
# Iterate across all loci/sample pairs.
# Handle/record lack of hits/sequences. 
# Check for really dodgy sequences. Differences in trimming?
# 	range of allele sizes. 
# set up directories to preserve results - fasta seqs dbs, alignments, summarytables.
# Log of files already processed as MD5_Checksum?
# Work RUN/PROJECT IDs. Could be used to test if a run has already been analysed. 
# Stats on variant frequencies
# Genotyping and genotyping parameters. (use "method" on a common structure so new methods can be updated)
# 	Assign alleles as iterate through marker/samples (i.e. ignore info on appearance of seqs across runs.)? 
# 	Or process all samples then go back and choose alleles and extract.
# allow use of BLAST+ 
# Test for hits to multiple MIDs (and remove). Limit valid hits to within X bp of start?
# Specify design of amplicons: number and type of MIDs, number and type of primers. Other seqs. 
# Sort out whether to use primer hits in assigning markers? It currently IS used indirectely by selection of subsequence with good hits to the correct primers.
# Fix warning from fastacmd caused by submission without accession list. MIGHT NOW BE FIXED
# make each marker run return an object. Good structure to run in parallel.
# SPEED: Do one massive alignment per marker? Might save time. Oringinal seqs with counts and aligned subseqs can be re-merged using orginal seq as name in sample/marker specific tables. 
# 	what if dodgy seqs in the mix? Will the alignment be any good?
# 	Might be worth throwing known alleles into the mix!
# MUSCLE alignments with dodgy ends e.g. A_E3.MID-13 - left a lone C at start of alignment because end of primer is a bit variable amogst seqs. 
# 	how to quickly test for and fix this problem?  (proportion of ---- before and after subSeq. 
# Decide on whether to do everything in a project/run directory. Blast DBs are easiest if located in working directory. 


### RECENTLY DONE



#################### Functions

getTopBlastHits <- function(blastTableFile)  {		# returns the first hit for each query in the table. May now be partially redundant if selecting for number of blast hits returned..
	blastResults <- read.delim(blastTableFile, header=F)
	names(blastResults) <- c("query", "subject", "percentId", "aliLength", "mismatches", "gapOpenings", "q.start","q.end", "s.start","s.end", "p_value", "e_value")
	topHits <- blastResults[match(unique(blastResults$query), blastResults$query),]
}


setUpBlastDb <- function(inputFastaFile, formatdbPath, blastdbName, indexDb="F")  {
	formatdbCommand <- paste(formatdbPath, "-i", inputFastaFile ,  "-p F -o", indexDb ,"-n", blastdbName)
	system(formatdbCommand)
}

getSubSeqs <- function(testPairSeqList, fTopHits, rTopHits, bothPrimerHits)  {
	seqGroup <- list()
	listWithPrimerHits <- intersect(testPairSeqList , bothPrimerHits)
	queryList <- paste(listWithPrimerHits, collapse=",")

	fShort <- fTopHits[match(listWithPrimerHits ,fTopHits$query),]
	rShort <- rTopHits[match(listWithPrimerHits ,rTopHits$query),]
	# could put check in here that all subject == thisMarker.

	primerHitTable <- merge(fShort, rShort, by=c("query", "subject"),suffixes = c(".f",".r"))
	primerHitTable$strand <- ifelse(primerHitTable$q.start.f  > primerHitTable$q.start.r, 2,1)
	primerHitTable$subStart <- ifelse(primerHitTable$strand == 1, primerHitTable$q.end.f,primerHitTable$q.end.r)
	primerHitTable$subEnd <- ifelse(primerHitTable$strand == 2, primerHitTable$q.start.f,primerHitTable$q.start.r)

	listWithBothGoodHits <- primerHitTable$query
	if(length(listWithBothGoodHits) < 1) {	# no remaining sequence IDs. Do not run fastacmd.
		return(seqGroup)
	}
 	# pull all these seqs from the blastDb together.
	queryList <- paste(listWithBothGoodHits , collapse=",")
	fastacmdCommand  <- paste(fastacmdPath, "-p F -t T -d", "inputSeqs" , "-o", subseqFileName,  "-s", queryList)
	system(fastacmdCommand)
	#seqGroup[[seqId]] <- read.fasta(subseqFileName, as.string=T)
	tempSeqGroup <- read.fasta(subseqFileName, as.string=T)
	# returned with "lcl|" prefix
	names(tempSeqGroup ) <-  sub("lcl\\|", "", names(tempSeqGroup ))


	# for each seq, extract the relevant portion, reverse if strand = 2
	for(seqId in listWithBothGoodHits)  {
		tableRow <- match(seqId, primerHitTable$query)
		strand <- primerHitTable$strand[tableRow]
		subStart <- primerHitTable$subStart[tableRow]
		subEnd <- primerHitTable$subEnd[tableRow]
		seq <- unlist(getSequence(tempSeqGroup[seqId], as.string=T))
		if(strand == 2)  {
			seq <- c2s(rev(comp(s2c((seq)))))
		}
		seqGroup[seqId] <- seq 
	}
	return(seqGroup)
}






####################

# seqinr http://cran.r-project.org/web/packages/seqinr/index.html
library(seqinr)


# system info
#formatdbPath
#blastAllPath
#musclePath
#params
#C:\Temp
# need to use correct system delimiter "\\" (Windows) or "/" (Windows or Unix)
formatdbPath <- "C:/Users/Public/Apps/Blast/bin/formatdb.exe"
fastacmdPath <- "C:/Users/Public/Apps/Blast/bin/fastacmd.exe"
blastAllPath <- "C:/Users/Public/Apps/Blast/bin/blastall.exe"
musclePath <- "C:/Users/Public/Apps/Muscle/muscle3.8.31_i86win32.exe"

#sysSep <- ifelse

###################

### set paramters. 

analysisDir <-  "C:/Users/dave/HalfStarted/mlgt"
setwd( analysisDir )
projectName <- "testProject"
runName <- "testRun"

# load  marker sequences

# load sample barcodes/MIDs

sampleFile <- "samples_ALL.tab"
sampleTable <- read.delim(sampleFile, sep=";")

### test and create directories for this project/run.
overWriteBaseData <- FALSE
runPath <- paste(analysisDir, projectName, runName, sep="/")
if(file.exists(runPath))  {
	overWriteBaseData <- readline("A folder for this run already exists, do you want to write over this data? (")
} else {
	dir.create(runPath, recursive=T)
}




## could easy make this a function
## need to slim down to unique pairs (e.g. once for each MID)
fastaFile <- paste(runPath,"fTags.fasta", sep="/")
tempTable <- unique(subset(sampleTable, select=c(tag_f, sample_name)))
write.fasta(as.list(tempTable$tag_f), tempTable$sample_name, file=fastaFile )
#blastdbName <- "fTags"
#formatdbCommand <- paste(formatdbPath, "-i", fastaFile,  "-p F -n", blastdbName)
#system(formatdbCommand)
setUpBlastDb(fastaFile , formatdbPath, blastdbName="fTags")


fastaFile <- "rTags.fasta"
tempTable <- unique(subset(sampleTable, select=c(tag_r, sample_name)))
write.fasta(as.list(tempTable$tag_r), tempTable$sample_name, file=fastaFile )
blastdbName <- "rTags"
formatdbCommand <- paste(formatdbPath, "-i", fastaFile,  "-p F -n", blastdbName)
system(formatdbCommand)


### primers as blast DB
fastaFile <- "fPrimers.fasta"
tempTable <- unique(subset(sampleTable, select=c(primer_f, marker)))
write.fasta(as.list(tempTable$primer_f), tempTable$marker, file=fastaFile )
blastdbName <- "fPrimers"
formatdbCommand <- paste(formatdbPath, "-i", fastaFile,  "-p F -n", blastdbName)
system(formatdbCommand)

fastaFile <- "rPrimers.fasta"
tempTable <- unique(subset(sampleTable, select=c(primer_r, marker)))
write.fasta(as.list(tempTable$primer_r), tempTable$marker, file=fastaFile )
blastdbName <- "rPrimers"
formatdbCommand <- paste(formatdbPath, "-i", fastaFile,  "-p F -n", blastdbName)
system(formatdbCommand)

## markers as blast DB
fastaFile <- "HLA_MARKERS.fasta"
blastdbName <- "markerSeqs"
formatdbCommand <- paste(formatdbPath, "-i", fastaFile,  "-p F -n", blastdbName)
system(formatdbCommand)


## input as blast DB with index (useful for fast sub-sequence retrieval?)
inputFastaFile <- "3.TCA.454Reads.fna"		# NCL Run_756948
#blastdbName <- "inputSeqs"
#formatdbCommand <- paste(formatdbPath, "-i", inputFastaFile ,  "-p F -o T -n", blastdbName)	# indexed
#system(formatdbCommand)
setUpBlastDb(inputFastaFile, formatdbPath, blastdbName="inputSeqs", indexDb="T") 


#### blasting input seqs against markers, primers and MIDs.

#write.fasta(as.list(sampleTable$tag_r), sampleTable$sample_name, file="rTags.fasta")

fastaFile <- "fewReads.fasta"
fastaFile <- "moreTestSeqs.fasta"
fastaFile <- inputFastaFile 

## Blast against MIDs. reduce blast word size accordingly. Need exact match for full length or forward and reverse tags.

## If f and r tags are identical, don't need to do both?
#dbPath <- paste(getwd(), "rTags", sep)
rMinTagSize <- 10	# limit blast hits to perfect matches.
dbName <- "rTags"
blastOutFile <- "blastOut.rTags.tab"
blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile , "-W", rMinTagSize  ,"-m 8 -b 2 -S 3 -o", blastOutFile )
#blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile , "-W", rMinTagSize  ,"-m 8 -o", blastOutFile )	# working
system(blastCommand )

fMinTagSize <- 10	# limit blast hits to perfect matches.
dbName <- "fTags"
blastOutFile <- "blastOut.fTags.tab"
blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile , "-W", fMinTagSize ,"-m 8 -b 2 -S 3 -o", blastOutFile )
system(blastCommand )


## blast against primers.
#fastaFile <- "fewReads.fasta"
dbName <- "rPrimers"
blastOutFileName <- "blastOut.rPrimers.tab"
blastOutFile <- paste(runPath, blastOutFileName, sep="/")
blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile, "-W", 11 ,"-m 8 -b 1 -o", blastOutFile )
#blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile, "-W", 11 ,"-m 8 -o", blastOutFile )
system(blastCommand )

#fastaFile <- "fewReads.fasta"
dbName <- "fPrimers"
blastOutFileName <- "blastOut.fPrimers.tab"
blastOutFile <- paste(runPath, blastOutFileName, sep="/")
blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile, "-W", 11 ,"-v", 4 , "-m 8 -b 1  -o", blastOutFile )
system(blastCommand )

## blast against markers
#fastaFile <- "fewReads.fasta"
dbName <- "markerSeqs"
blastOutFileName <- "blastOut.markers.tab"
blastOutFile <- paste(runPath, blastOutFileName, sep="/")
# -b to restrict reported hits to 1. ### Could put hit quality filter in here or later. 
blastCommand <- paste(blastAllPath , "-p blastn -d", dbName , "-i", fastaFile, "-W", 11 , "-m 8 -b 1 -o", blastOutFile )		
system(blastCommand )



## is blast 'v' useful. Trying to reduce number of reported hits. Only need top 2 or even less. 

## Need to make DB of marker sequences. Use primers too?  


# load run data (one seq at a time?). And store efficiently?
# there may be some identical sequences (even with random ends?). 



# assign sequences to samples
# 100% full length match to a single MID (or combination of MIDs if different ends are used)

##read blast result table.


# assign sequneces to markers. 
# Top BLAST hit with minimum level of identity
#topHits <- blastResults[match(unique(blastResults$query), blastResults$query),]
topHits <- getTopBlastHits("blastOut.markers.tab")		# redundant method if only storing top hit for each anyway.
markerMap <- split(as.character(topHits[,1]), topHits[,2])
## TO BE REPLACED BY:
	topHits <- getTopBlastHits("blastOut.markers.tab")
	topHits$strand <- ifelse(topHits$s.end > topHits$s.start, 1,2)
	fMarkerMap <- split(as.character(topHits$query[topHits$strand==1]), topHits$subject[topHits$strand==1])
	rMarkerMap <- split(as.character(topHits$query[topHits$strand==2]), topHits$subject[topHits$strand==2])



# read results from samples and make marker map. 

#blastResults <- read.delim("blastOut.rTags.tab", header=F)
#names(blastResults) <- c("query", "subject", "percentId", "aliLength", "mismatches", "gapOpenings", "q.start","q.end", "s.start","s.end", "p_value", "e_value")
#topHits <- blastResults[match(unique(blastResults$query), blastResults$query),]


topHits <- getTopBlastHits("blastOut.rTags.tab")
sampleMap <- split(as.character(topHits[,1]), topHits[,2])
## NEED TO MAKE SAMPLEMAP WITH HITS TO MID IN BOTH FORWARD AND REVERSE STRANDS like marker hits are split.
## Requires retention of 2 blast hits per sequence.
	topSampleHits <- getTopBlastHits("blastOut.rTags.tab")
	topSampleHits <- read.delim("blastOut.rTags.tab", header=F)
	names(topSampleHits ) <- c("query", "subject", "percentId", "aliLength", "mismatches", "gapOpenings", "q.start","q.end", "s.start","s.end", "p_value", "e_value")
	topSampleHits$strand <- ifelse(topSampleHits$s.end > topSampleHits$s.start, 1,2)
	fSampleMap <- split(as.character(topSampleHits$query[topSampleHits$strand==1]), topSampleHits$subject[topSampleHits$strand==1])
	rSampleMap <- split(as.character(topSampleHits$query[topSampleHits$strand==2]), topSampleHits$subject[topSampleHits$strand==2])
	# combind sampleMaps to give sequences with MIDs in both orientations.
	pairedSampleMap <- lapply(names(fSampleMap), FUN=function(x) intersect(fSampleMap[[x]], rSampleMap[[x]]))
	names(pairedSampleMap) <- names(fSampleMap)
	# test this worked correctly:  intersect(pairedSampleMap[[thisSample]], fSampleMap[[thisSample]])
	
#sampleMapClean <- intersect(fSampleMap, rSampleMap)


#integrate marker map and sample map to get fully assigned sequences. 
#testPairSeqList <- intersect(sampleMap[["MID-24"]], markerMap[["A_E2"]])

# Remove primer sequences and MIDs? Primers must be closest to amplicon. 
# In our study, primers are degenerate, the sequence corresponding to the primer is unreliable. 

# TO BE TAKEN OUT
fTopHits <- getTopBlastHits("blastOut.fPrimers.tab")	# is this a redundant method?
rTopHits <- getTopBlastHits("blastOut.rPrimers.tab")

bothPrimerHits <- intersect(fTopHits$query,rTopHits$query) # (almost all)

# subsequence required is marked by blast hits. But start/ end may either be in r or f hits. 
# sequences may fail if they don't have perfect hits in both.
# TO DO Might be best to do a merge on the two primer hit tables.
#primerCoords <- unlist((c(fTopHits[i,c("q.start","q.end")], rTopHits[i,c("q.start","q.end")])))
#strand <- ifelse(primerCoords[3] < primerCoords[2], 2,1)	# 2 -ve strand
#primerCoords <- sort(primerCoords)		# replace with tapply
#subCoords <- c(primerCoords[2]+1, primerCoords[3]-1)

#### is it best to extract full sequence then make subsequence or extract subsequence straighaway.

#retSeq <- as.character(fTopHits$query[i])
subseqFileName <- paste(runPath, "subseqFile.fasta",sep="/")

#fastacmdCommand  <- paste(fastacmdPath, "-p F -d", blastdbName , "-s", retSeq , "-L", paste(subCoords[1],subCoords[2],sep=","))
#fastacmdCommand  <- paste(fastacmdPath, "-p F -d", blastdbName , "-o", subseqFileName,  "-s", retSeq , "-L", paste(subCoords[1],subCoords[2],sep=","), "-S", strand)
#system(fastacmdCommand)
#retSeqObject <- read.fasta(file=stdin(system(fastacmdCommand)))



#seqGroup[i] <- read.fasta(subseqFileName, as.string=T)
#subseqFileName <- "subseqFile.fasta"
minAmpliconLength <- 40

#########################################
#### iterations
##########################################

markerSampleList <- list()
runSummaryTable <- data.frame()
alleleDb <- list()


for(thisMarker in names(markerMap)) {
#for(thisMarker in names(markerMap)[16:19]) {	# temp to finish off

print(thisMarker)
#thisMarker <- "DQA1_E2"

## might need to combine all these to return a single item.
summaryList <- list()
summaryTable <- data.frame()
markerSequenceCount <- list("noSeq"=0)		#  BUG? requires some data otherwise won't sum properly with localSequenceCount.
alleleList <- list() 
variantList <- list()
alleleCount <- 1


for(thisSample in names(sampleMap)) {
	#print(thisSample)

	
	#testPairSeqList <- intersect(pairedSampleMap[[thisSample]], markerMap[[thisMarker]])
	testPairSeqList <- intersect(sampleMap[[thisSample]], markerMap[[thisMarker]])


seqTable <- data.frame()
localAlleleNames <- c("NA","NA","NA")
localAlleleFreqs <- c(0,0,0)

## go through all seq's mapped to this marker/sample pair.
## extract the corresponding sequence delimited by the top blast hits on the primers.  IS THIS THE BEST WAY?
##		Simple improvement: minimum blast hit length to primer to keep. 

## internal Function

recordNoSeqs <- function(summaryTable)  {		# to record no seqs before skipping out. 
		summaryRow <- data.frame(marker=thisMarker, sample=thisSample, numbSeqs=0,numbVars=0,
			varName.1="NA", varFreq.1= 0,
			varName.2="NA", varFreq.2= 0,
			varName.3="NA", varFreq.3= 0)
		summaryTable <- rbind(summaryTable, summaryRow)
		return(summaryTable)
}



if(length(testPairSeqList) < 1) {
	#summaryList[[thisMarker]][[thisSample]] <- NA	
	summaryTable  <- recordNoSeqs(summaryTable)
	next ;	# skip to next sample
} 



seqGroup <- getSubSeqs(testPairSeqList, fTopHits, rTopHits, bothPrimerHits)


# if no sequences returned, nothing to process. 
if(length(seqGroup) < 1 )  {
	summaryTable  <- recordNoSeqs(summaryTable)
	#summaryList[[thisMarker]][[thisSample]] <- NA	
	next ;		# go to next sample.
}

# looks pretty good, some very long seqs though. 
#str(seqGroup[[1]])

seqTable <- cbind(as.character(unlist(seqGroup)), names(unlist(seqGroup)))



localSequenceMap <- split(seqTable[,2], seqTable[,1])

localSequenceCount <- lapply(localSequenceMap , length)  # list named by sequence with counts.
localSequenceCount <- localSequenceCount[order(as.numeric(localSequenceCount), decreasing=T)]


# Condense number of sequence and map seqs to variants.  
#lapply(localSequenceMap , length)
localVariants <- names(localSequenceCount)

localVariantFileName <- paste("thisRun", thisMarker, thisSample, "variants.fasta",sep=".")  #"localVariants.fasta"
localVariantFile <- paste(runPath, localVariantFileName, sep="/")
write.fasta(as.list(localVariants) ,localVariants,file=localVariantFile ,open="w")

## test if variants are novel. 
## Give allele names?  
## Do with first three for now. 


alToRecord <- min(3,length(localSequenceCount))
if(alToRecord > 0)  {
	for (a in 1:alToRecord )  {
		if(is.null(variantList[[names(localSequenceCount[a])]]))  {    	# novel
			alleleName <- paste(thisMarker, alleleCount,sep=".")	
			variantList[[names(localSequenceCount[a])]] <- alleleName
			localAlleleNames[a] <- alleleName 
			localAlleleFreqs[a] <- localSequenceCount[[a]]
			alleleCount <- alleleCount + 1
		} else  {										# pre-existing alllele
			localAlleleNames[a] <- variantList[[names(localSequenceCount[a])]]
			localAlleleFreqs[a] <- localSequenceCount[[a]]		
		}
	}
}
# align sequences and store alignment.  # Is this needed for genotyping?  Maybe if some small differences in common seqs. Yes, if want to correct the sequences.
# Will need way to nicely visualise alignments. 

localAlignFileName <- paste("thisRun", thisMarker, thisSample, "align.fasta",sep=".")  #"localAlign.fasta"
localAlignFile <- paste(runPath, localAlignFileName, sep="/")
muscleCommand <- paste(musclePath, "-in", localVariantFile , "-out", localAlignFile , "-diags -quiet" )
system(muscleCommand)
# can read alignment in.
#localAlignment <- read.alignment(localAlignFile , format="fasta")


# sequence correction?  


# compile stats

if(length(localSequenceCount) >0 )  {	# cannot allow assignment from empty list as messes up class of list for remaining iterations
	summaryList[[thisMarker]] <- list()
	summaryList[[thisMarker]][[thisSample]] <- localSequenceCount 
}
summaryRow <- data.frame(marker=thisMarker, sample=thisSample, numbSeqs=nrow(seqTable),numbVars=length(localSequenceCount),
			varName.1=localAlleleNames[1], varFreq.1= localAlleleFreqs[1],
			varName.2=localAlleleNames[2], varFreq.2= localAlleleFreqs[2],
			varName.3=localAlleleNames[3], varFreq.3= localAlleleFreqs[3])
summaryTable <- rbind(summaryTable, summaryRow)

#sequence count across samples? 
# need to sum from summaryTable or from summaryList.
#markerSequenceCount <- 
#as.list(colSums(merge(m, n, all = TRUE), na.rm = TRUE))  # not working
markerSequenceCount   <- as.list(colSums(merge(markerSequenceCount  , localSequenceCount,  all = TRUE), na.rm = TRUE))
# might need to instantiate the markerSequenceCount if empty. 



}  # end of sample loop

markerSampleList[[thisMarker]] <- summaryTable
runSummaryRow <- data.frame(marker=thisMarker, assigneSeqs=sum(summaryTable$numbSeqs), assignedVariants=sum(summaryTable$numbVars), 
				minVariantLength=min(nchar(names(markerSequenceCount))), 
				maxVariantLength=max(nchar(names(markerSequenceCount))),
				minAlleleLength=min(nchar(names(variantList))), maxAlleleLength=max(nchar(names(variantList))))
runSummaryTable <- rbind(runSummaryTable, runSummaryRow)
if(length(variantList) > 0)  {
	alleleDb[[thisMarker]] <- variantList   # LATER: separate lists for alleles and variants?
}
}  # end of marker loop

####################################


# save.image("C:\\Users\\dave\\HalfStarted\\mlgt\\testProjectData.RData")

####################################



# http://stackoverflow.com/questions/5783241/variable-names-are-limited-to-256-bytes
# Running under R 2.13

length(variantList)
min(nchar(names(variantList)))		# shortest of variants retained as potential alleles
max(nchar(names(variantList)))		# longest of variants retained as potential alleles

min(nchar(names(markerSequenceCount)))	# shortest extracted variant
max(nchar(names(markerSequenceCount)))	# longest extracted variant

length(markerSequenceCount )
nrow(summaryTable)
sum(summaryTable$numbVars)
sum(summaryTable$numbSeqs)



addToSeqCount <- function(oldCount, newCount)  {
	

}


# How to combine results from one pair across all pairs or within a locus. Sum/merge the lists?






# choose alleles if possible


# output genotypes with alleles.



# match allleles to allele database. 



stopifnot(FALSE)
################DEVELOPMENT


### could be failing because merge collapses the results when the two list are identical to begin with.
markerSequenceCount <- list("noSeq"=0, "diff"=5)	# nope still not working. Adds the new names but doesn't summ the common values. 
markerSequenceCount   <- as.list(colSums(merge( localSequenceCount, markerSequenceCount  , all = TRUE), na.rm = TRUE))
newSequenceCount <- localSequenceCount
as.list(colSums(merge( localSequenceCount, markerSequenceCount  , all = TRUE), na.rm = TRUE))

markerSequenceCount   <- as.list(colSums(merge( localSequenceCount, markerSequenceCount  , all = TRUE), na.rm = TRUE))

markerSequenceCount <- lapply(markerSequenceCount, as.integer)
str(markerSequenceCount)
str(localSequenceCount)
localSequenceCount<- lapply(localSequenceCount, as.numeric)


markerSequenceCount   <- as.list(colSums(merge( localSequenceCount, markerSequenceCount  , all = TRUE), na.rm = TRUE))

localSequenceCount + markerSequenceCount  

localSequenceCounts <- markerSequenceCount   
#is this just going to be easier with a data.frame?

localSequenceCount["diff"] <- 2


list1 <- list("x"=16, "y"=2)
list2 <- list("x"=14, "y"=7)

